scripts/fetch_olean_cache.sh

```diff
@@ -3,6 +3,20 @@ set -x
 
 archive_url="https://oleanstorage.azureedge.net/mathlib/"
 
-new_git_sha=$(python scripts/look_up_olean_hash.py $lean_file_hash) || exit 0
-echo "equivalent Git sha: $new_git_sha"
+if new_git_sha=$(python scripts/look_up_olean_hash.py $lean_file_hash); then
+  echo "equivalent Git sha: $new_git_sha"
+else
+  # MAX_DEPTH is set in .github/workflows/build.yml
+  for new_git_sha in $(git log -$MAX_DEPTH --first-parent --pretty=format:%H)
+  do
+    if curl -sfI "$archive_url$new_git_sha.tar.gz" ; then
+      echo "found ancestor Git sha: $new_git_sha"
+      break
+    fi
+    new_git_sha=""
+  done
+fi
+# exit if there were no successful requests
+[ "$new_git_sha" != "" ] || exit 0
+
 curl "$archive_url$new_git_sha.tar.gz" | tar xz src
```

Comment:
It can be the case that some `*.lean` files have been deleted between `$new_git_sha` and the current revision.  That is, we would have some `*.olean` files without corresponding `*.lean` files.  These need to be removed.  Otherwise, they would continue to live forever in the tarballs.

Reply:
Ah, good catch. I'll copy some lines from the old `purge_olean.sh` which was deleted when we moved to Github actions in #1893.

Reply:
Wow, I just realized that the curl tar line is overwriting all `.lean` files as well! Whoops.

Reply:
OK, I did a test where I deleted `surreal.lean` and then downloaded the result [from this builld](https://github.com/leanprover-community/mathlib/runs/543413565#step:9:1). The cache file didn't contain surreal.lean or surreal.olean, so I think both issues here are addressed.